ui-speedspacechart/src/components/utils.ts

```diff
@@ -266,24 +266,49 @@ export const createSvgBlobUrl = (svgString: string): string => {
   return URL.createObjectURL(blob);
 };
 
-export const findPreviousAndNextPosition = <T = string | number>(
-  data: LayerData<T>[],
-  cursorX: number,
-  xPositionReference: (ref: number) => number
-) => {
-  const previousPosition = data.findLast(
-    ({ position }) => xPositionReference(position.start) <= cursorX!
-  );
-  const nextPosition = data.find(({ position }) => xPositionReference(position.start) >= cursorX!);
+/** Retrieve the index of an element in the list.
+ * If the element is not found, return the predecessor.
+ * @param data - The **sorted** data array.
+ * @param element - The element to search
+ * @param lambda - The function to extract the value from the data list.
+ */
+export const binarySearch = <T>(data: T[], element: number, lambda: (element: T) => number) => {
+  let left = 0;
+  let right = data.length;
+  while (left < right) {
+    const m = Math.floor((left + right) / 2);
+    if (lambda(data[m]) < element) {
+      left = m + 1;
+    } else {
+      right = m;
+    }
+  }
+
+  if (element !== lambda(data[left])) {
+    left--;
+  }
 
-  return { previousPosition, nextPosition };
+  return left;
 };
 
-export const getStopPosition = (
-  position: LayerData<string>['position'],
+/** Transform a position in km into a position on the x-axis in pixels */
+export const positionToPosX = (
+  position: number,
+  maxPosition: number,
   width: number,
   ratioX: number,
-  maxPosition: number
+  leftOffset: number = 0
 ) =>
-  position.start * ((width - MARGINS.CURVE_MARGIN_SIDES) / maxPosition) * ratioX +
-  MARGINS.CURVE_MARGIN_SIDES / 2;
+  (position / maxPosition) *
+    ((width - MARGINS.MARGIN_LEFT - MARGINS.MARGIN_RIGHT - MARGINS.CURVE_MARGIN_SIDES) * ratioX) +
+  MARGINS.CURVE_MARGIN_SIDES / 2 +
+  MARGINS.MARGIN_LEFT +
+  leftOffset;
```

Comment:
Maybe it would be easier to read like this
```suggestion
  const xWidth = width - MARGINS.MARGIN_LEFT - MARGINS.MARGIN_RIGHT - MARGINS.CURVE_MARGIN_SIDES;
  const leftMargin = MARGINS.CURVE_MARGIN_SIDES / 2 + MARGINS.MARGIN_LEFT + leftOffset;
  return (position / maxPosition) * (xWidth * ratioX) + leftMargin;
```